library(tidyverse)
library(plyr)
library(stringr)
library(rehh)
library(tibble)
###Input for this script are phased haplotypes and the markers under selection generated by running "R_script_rehh"
#Note that this script might be somewhat difficult to read. I advice you to not run the for loops initially and run the lines one by one.
#To be able to run the code line by line and assess the generated datafiles individually, 
#set the variable "num" as a chromosome number and "word" as a marker on that chromosome

setwd("~/Documents/Extra_GRSM/Outlying_GRSM")
range<-read.table("227_ihs_markers",header=T) #note that this is not 233 markers. I merged the GRSM SNPs with the outlying SNPs, which loses some SNPs 
num<-unique(range$CHR)
for(number in num){
  #get family information
  setwd("~/Documents/Extra_GRSM/Outlying_GRSM/Phasing")
  famname<-paste0("GRSM_Outlying_Merged_chr",number,".fam")  
  fam1 <- read.table(famname)
  fam1$MERGE <- paste0(fam1$V2,"_1")
  fam2 <- read.table(famname)
  fam2$MERGE <- paste0(fam2$V2,"_2")
  fam <- rbind(fam1,fam2)
  fam <- fam %>% select(V1,V2,MERGE) #This df is necessary to represent the two chromosomes in one individual
  bimname<-paste0("GRSM_Outlying_Merged_chr",number,".bim")  
  bim<-read.table(bimname)
  #get haplotypes
  setwd("~/Documents/Extra_GRSM/Outlying_GRSM/Sweep")
  vcfname<-paste0("VCF_GRSM+Outlying_chr",number,".gz")  
  vcf <- data2haplohh(hap_file = vcfname,polarize_vcf = FALSE,min_perc_geno.mrk = 50,
                      vcf_reader = "vcfR", verbose = FALSE)
  haplo<-vcf@haplo
  haplo<-as.data.frame(haplo)
  #for loop
  setwd("~/Documents/Extra_GRSM/Outlying_GRSM")
  range<-read.table("227_ihs_markers",header = T)
  range<-subset(range, CHR==number)
  range<-range$MARKER
  range<-as.character(range)
  for(word in range){#lines 39-50 generate a region of 41 SNPs with the middle SNP being a marker under selection. 
                     #This region is 'eroded' (i.e. removing SNPs) from the edges until the size of the region/haplotype is approximate 0.5 Mb
    haplosmall <- haplo[,c((which( colnames(haplo)==word)-20):(which( colnames(haplo)==word)+20))] 
    alleles<-colnames(haplosmall)
    alleles<-as.data.frame(alleles)
    colnames(alleles)<-"V2"
    alleles<-merge(bim,alleles,by="V2")
    alleles<-arrange(alleles,V4)
    for(i in c(1:40)){ #For loop to decrease length haplotype to 0.5 Mb
      alleles<-if(max(alleles$V4)-min(alleles$V4)<=0.5*10^6) alleles else alleles[-1,]
      alleles<-if(max(alleles$V4)-min(alleles$V4)<=0.5*10^6) alleles else alleles[-nrow(alleles),]
    }
    alleles<-as.character(alleles$V2)
    haplosmall<-haplosmall[,alleles]
    haplosmall <- haplosmall %>% unite(alleles,sep="") #a dataset with the phased states of several markers in one column per chromsome in your population
    dplydf <- ddply(haplosmall,1:ncol(haplosmall),nrow) #Dataset with occurrence of specific haplotypes on all chromosomes in your population
    df <- tibble::rownames_to_column(haplosmall, "MERGE")
    df <-merge(df,fam, by="MERGE")
    df$MERGE<-df$V1
    df$V1<-NULL
    df$V2<-NULL
    df$MERGE<-as.character(df$MERGE) #df is a dataframe including the family identifier and the haplotypes of around 0.5 Mb
    
    dfGRSM<-subset(df,MERGE=="GRSM")
    dfGRSM<-ddply(dfGRSM,1:ncol(dfGRSM),nrow) #count haplotypes in GRSM
    f <- function(vec) {#this function add the count data of GRSM to df
      if(length(.dfGRSM <- which(vec == dfGRSM$alleles))) .dfGRSM else NA
    }
    dplydf$GRSM <- dfGRSM$V1[mapply(f,dplydf$alleles)] 
    
    dfTN<-subset(df,MERGE=="TN")
    dfTN<-ddply(dfTN,1:ncol(dfTN),nrow)
    f <- function(vec) {
      if(length(.dfTN <- which(vec == dfTN$alleles))) .dfTN else NA
    }
    dplydf$TN <- dfTN$V1[mapply(f,dplydf$alleles)]
    
    dfSC<-subset(df,MERGE=="SC")
    dfSC<-ddply(dfSC,1:ncol(dfSC),nrow)
    f <- function(vec) {
      if(length(.dfSC <- which(vec == dfSC$alleles))) .dfSC else NA
    }
    dplydf$SC <- dfSC$V1[mapply(f,dplydf$alleles)]
    
    dfNC<-subset(df,MERGE=="NC")
    dfNC<-ddply(dfNC,1:ncol(dfNC),nrow)
    f <- function(vec) {
      if(length(.dfNC <- which(vec == dfNC$alleles))) .dfNC else NA
    }
    dplydf$NC <- dfNC$V1[mapply(f,dplydf$alleles)]
    
    dfWV<-subset(df,MERGE=="WV")
    dfWV<-ddply(dfWV,1:ncol(dfWV),nrow)
    f <- function(vec) {
      if(length(.dfWV <- which(vec == dfWV$alleles))) .dfWV else NA
    }
    dplydf$WV <- dfWV$V1[mapply(f,dplydf$alleles)]
    
    dplydf[is.na(dplydf)] <- 0 #dplydf now includes per haplotype the absolute count per population
    dplydf$hapname<-"Hap"
    hapnum<-c(1:length(dplydf$V1))
    dplydf$hapname<-paste0(dplydf$hapname,"_",hapnum) #hapnum is an identifier of the haplotype
    
    smalldf <- subset(dplydf,GRSM>=0) 
    #smalldf$hapname<-str_pad(smalldf$hapname, 10, side="right", pad=" ") #this code is useful for alignment in some cases
    smalldf1<- smalldf%>% select(hapname,alleles)
    name<-paste0("Chr",number,"_",word,"_Haplotypes")
    setwd("~/Documents/Extra_GRSM/Outlying_GRSM/Haplotype_inspection_227Markers/Summary")
    name<-paste0("Chr",number,"_Summary_",word,"_Haplotypes.txt")
    write.table(smalldf,file=name,quote=F,sep="\t",row.names = F) #writes table with absolute haplotype counts
    smalldf2<-subset(dplydf,GRSM>=0)
    #smalldf2$hapname<-str_pad(smalldf2$hapname, 10, side="right", pad=" ")
    smalldf2$GRSM<-smalldf2$GRSM/(sum(smalldf2$GRSM))
    smalldf2$TN<-smalldf2$TN/(sum(smalldf2$TN))
    smalldf2$WV<-smalldf2$WV/(sum(smalldf2$WV))
    smalldf2$NC<-smalldf2$NC/(sum(smalldf2$NC))
    smalldf2$SC<-smalldf2$SC/(sum(smalldf2$SC))
    setwd("~/Documents/Extra_GRSM/Outlying_GRSM/Haplotype_inspection_227Markers/Frequency")
    name<-paste0("Chr",number,"_frequency_",word,"_Haplotypes.txt")
    write.table(smalldf2,file=name,quote=F,sep="\t",row.names = F) #writes table with frequency of haplotype counts
    
  }
}

##Generate haplotype files with only haplotype present in the GRSM
setwd("~/Documents/Extra_GRSM/Outlying_GRSM")
range<-read.table("227_ihs_markers",header=T)
num<-unique(range$CHR)
for(number in num){
  setwd("~/Documents/Extra_GRSM/Outlying_GRSM")
  range<-read.table("227_ihs_markers",header=T)
  range<-subset(range, CHR==number)
  range<-range$MARKER
  range<-as.character(range)
  for(word in range){
    namedf2<-paste0("Chr",number,"_frequency_",word,"_Haplotypes.txt")
    setwd("~/Documents/Extra_GRSM/Outlying_GRSM/Haplotype_inspection_227Markers/Frequency")
    df2<-read.table(namedf2,header=T, colClasses = c("character",rep("numeric",6),"character"))
    df2<-subset(df2,GRSM>0)
    df2$GRSM<-formatC(df2$GRSM, digits = 4 , format = "f")
    df2$TN<-formatC(df2$TN, digits = 4 , format = "f")
    df2$SC<-formatC(df2$SC, digits = 4 , format = "f")
    df2$NC<-formatC(df2$NC, digits = 4 , format = "f")
    df2$WV<-formatC(df2$WV, digits = 4 , format = "f")
    setwd("~/Documents/Extra_GRSM/Outlying_GRSM/Haplotype_inspection_227Markers/Frequency_GRSM")
    write.table(df2,namedf2,quote=F,row.names=F,col.names = T,sep="\t")
  }
}
